### PR TITLE
Automatically resize charts on updates.

### PR DESCRIPTION
**Pull request by domoritz**
cc @yifanwu

### CODE REVIEW

**Comment by jheer** [comment]
A word of warning (just to be safe): setting `resize: true` can cause layout to re-run on *every view update*, not just streaming data updates. That means if an interactive selection changes the axis labels and that changes the chart size, everything might "jump" between frames. That is one reason why resize is not true by default (by default a layout "resize" is only run when the chart width and/or height signals are modified). The other reason is for performance considerations, as "fit" and "pad" auto-sizing require running *two* pulse propagation cycles. 

If instead you'd like a way to trigger re-layout from the View API, that would not be hard to add. Just file a feature request issue on the vega-view repo.

**Comment by kanitw** [comment]
@domoritz Sounds like we shouldn't merge this then.

**Comment by domoritz** [comment]
The problem is that the legend in the streaming example is cut off. At least, we need an option to enable automatic resizing.

**Comment by kanitw** [comment]
>  in the streaming example
which one?

**Comment by kanitw** [comment]
Also, for streaming use case, seems like re-layout trigger API that Jeff suggests is more appropriate?

>  an option to enable automatic resizing.
sounds more reasonable than enable by default.

**Comment by domoritz** [comment]
The streaming example is forthcoming. I filed an issue at https://github.com/vega/vega-view/issues/6. 

I still think we need an option to enable automatic resizing. I will add a top level option `resize`. 

Another option I thought about last night is to automatically detect whether a spec has named datasources. However, this is again only a heuristic.

**Comment by domoritz** [comment]
@jheer Can you confirm that Vega-Lite has to use `autosize: "pad"` because we use `layout`? I'm asking because I want to add a top level property `autosize` if we do support other types and only `resize` if not.

**Comment by jheer** [comment]
Yes, unless you somehow track all your space usage manually and incorporate that into width/height/padding signal calculations, you'll definitely need to use "pad".

**Comment by domoritz** [comment]
@kanitw I pushed another commit with a new top level option `autoResize`.

File: src/compile/compile.ts [hunk 1]
```diff
@@ -56,11 +56,18 @@ function assemble(model: Model, topLevelProperties: TopLevelProperties) {
...
+  // remove autoResize as it has to be put under autosize
+  const topLevelProps = extend({}, topLevelProperties);
+  delete topLevelProps.autoResize;
+
   const output = {
     $schema: 'http://vega.github.io/schema/vega/v3.0.json',
     ...(model.description ? {description: model.description} : {}),
-    autosize: 'pad', // By using Vega layout, we don't support custom autosize
-    ...topLevelProperties,
+    autosize: {
```
Comment (kanitw) [inline src/compile/compile.ts]: Wonder if you think
```
autosize: topLevelProperties.autoResize ? {type: 'pad', resize: true} : 'pad'
```
is a bit nicer, and won't require changing all existing specs.

Comment (domoritz) [inline src/compile/compile.ts]: Done

**Comment by domoritz** [comment]
Done.

File: src/compile/compile.ts [hunk 2]
```diff
@@ -56,11 +56,16 @@ function assemble(model: Model, topLevelProperties: TopLevelProperties) {
   // Config with Vega-Lite only config removed.
   const vgConfig = model.config ? stripConfig(model.config) : undefined;
 
+  // remove autoResize as it has to be put under autosize
+  const topLevelProps = extend({}, topLevelProperties);
```
Comment (kanitw) [inline src/compile/compile.ts]: `const {_autoResize, ...topLevelProps} = topLevelProperties;`  ?

Comment (domoritz) [inline src/compile/compile.ts]: does the _ work?

**Comment by domoritz** [comment]
Thanks!

**Pull request merged**